rule/sexpr/lexer.go

```diff
@@ -64,3 +73,385 @@ func isComment(r rune) bool {
+	s.charCount--
+	s.lineCharCount--
+	if rn == '\n' {
+		s.lineCount--
+		s.lineCharCount = s.previousLineCharCount
+		s.previousLineCharCount--
+	}
+
+}
+
+// scanWhitespace scans a contiguous sequence of whitespace
+// characters.  Note that this will consume newlines as it goes,
+// lexically speaking they're insignificant to the language.
+func (s *Scanner) scanWhitespace() (Token, string, error) {
+	var b bytes.Buffer
+	for {
+		rn, err := s.readRune()
+		if err != nil {
+			se := err.(*ScanError)
+			if se.EOF {
+				// We'll get EOF next time we try to
+				// read a rune anyway, so we don't
+				// have to care about it here, which
+				// simplifies things.
+				return WHITESPACE, b.String(), nil
+
+			}
+			return WHITESPACE, b.String(), err
+
+		}
+		if !isWhitespace(rn) {
+			s.unreadRune(rn)
+			break
+		}
+		b.WriteRune(rn)
+	}
+	return WHITESPACE, b.String(), nil
+}
+
+// scanString returns the contents of single, contiguous, double-quote delimited string constant.
+func (s *Scanner) scanString() (Token, string, error) {
+	var b bytes.Buffer
+	escape := false
+	for {
+		rn, err := s.readRune()
+		if err != nil {
+			se := err.(*ScanError)
+			if se.EOF {
+				// we reached the end of the file
+				// without seeing a terminator, that's
+				// an error.
+				return STRING, b.String(), s.newScanError("unterminated string constant")
+			}
+			return STRING, b.String(), err
+		}
+		if escape {
+			b.WriteRune(rn)
+			escape = false
+			continue
+		}
+		if isString(rn) {
+			break
+		}
+		escape = rn == '\\'
+		if !escape {
+			b.WriteRune(rn)
+		}
+	}
+	return STRING, b.String(), nil
+}
+
+// scanNumber scans a contiguous string representing a number.  As we
+// have to handle the negative numeric form, it's possible that the
+// '-' rune can prefix a number.  This is problematic because '-' can
+// also be a symbol referring to the arithmetic operation "minus" -
+// that confusion is resolved by scanNumber, and should it consider
+// the latter case to be true it will return a SYMBOL rather than a
+// NUMBER.
+func (s *Scanner) scanNumber() (Token, string, error) {
+	var b bytes.Buffer
+
+	// We can be certain this isn't EOF because we will already
+	// have read and unread the rune before arriving here.
+	rn, err := s.readRune()
+	if err != nil {
+		// Something drastic happened, because we read this fine the first time.
+		return NUMBER, "", err
+	}
+
+	// Whatever happens we'll want the rune.
+	b.WriteRune(rn)
+
+	// Deal with the first rune.  Numbers have special rules about
+	// the first rune, specifically it, and only it, may be the
+	// minus symbol. When we loop later any occurrence of '-' will
+	// be an error.
+	if rn == '-' {
+
+		// Now we look ahead to see if a number is coming, if
+		// its anything else then this isn't a negative
+		// number, but some other form.
+		rn, err := s.readRune()
+
+		if err != nil {
+			se := err.(*ScanError)
+			if se.EOF {
+				// In reality having '-' as the final
```

Comment:
Not sure to understand this comment. 
It can be everything else, why this comment is oriented to have a `-` here?

Reply:
Oh I got it, because we first read the `-` character, then we read again and we got an EOF error. So the last character is exactly a `-`.